File: libs/power-log-parser/src/lib/parsers/battlegrounds-battle-over-parser.ts
import { BlockType, CardIds, CardType, GameTag, PlayState, Zone } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Action, FullEntity, Node, NodeType, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsBattleOverParser implements ActionParser {
	readonly ParserName = 'BattlegroundsBattleOverParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, helper: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = helper;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			!this.GameState.BattleResultSent &&
			node.Type === NodeType.TagChange &&
			(node.Object as TagChange).Name === (GameTag.BOARD_VISUAL_STATE as number) &&
			(node.Object as TagChange).Value === 1
		);
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === NodeType.Action &&
			(node.Object as Action).Type === (BlockType.TRIGGER as number) &&
			(node.Object as Action).EffectIndex === 14
		);
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		this.GameState.BattleResultSent = true;
		const tagChange = node.Object as TagChange;
		let opponentCardId: string | null = this.GameState.BgsCurrentBattleOpponent;
		let opponentPlayerId = this.GameState.BgsCurrentBattleOpponentPlayerId;
		const mainPlayer = this.StateFacade.LocalPlayer!;
		if (opponentCardId == null || this.isBaconGhost(opponentCardId)) {
			const playerEntity = [...this.GameState.CurrentEntities.values()]
				.filter((entity) => entity.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number))
				.filter((entity) => entity.GetTag(GameTag.ZONE) === (Zone.PLAY as number))
				.filter((entity) => entity.GetEffectiveController() === mainPlayer.PlayerId)
				.filter(
					(entity) =>
						!entity.IsBaconBartender() &&
						!entity.IsBaconGhost() &&
						!entity.IsBaconEnchantment(),
				)
				.sort((a, b) => a.Id - b.Id)
				.pop();
			const nextOpponentPlayerId = playerEntity!.GetTag(GameTag.NEXT_OPPONENT_PLAYER_ID);

			const nextOpponentCandidates = [...this.GameState.CurrentEntities.values()]
				.filter((entity) => entity.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number))
				.filter((entity) => entity.GetTag(GameTag.PLAYER_ID) === nextOpponentPlayerId)
				.filter(
					(entity) =>
						!entity.IsBaconBartender() &&
						!entity.IsBaconGhost() &&
						!entity.IsBaconEnchantment(),
				);
			const nextOpponent = nextOpponentCandidates.length === 0 ? null : nextOpponentCandidates[0];

			opponentCardId = nextOpponent?.CardId ?? null;
			opponentPlayerId = nextOpponent?.GetTag(GameTag.PLAYER_ID) ?? 0;
		}

		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'BATTLEGROUNDS_BATTLE_RESULT',
				() => ({
					Type: 'BATTLEGROUNDS_BATTLE_RESULT',
					Value: {
						Opponent: opponentCardId,
						OpponentPlayerId: opponentPlayerId,
						Result: 'tied',
					},
				}),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const action = node.Object as Action;
		const actionEntity = this.GameState.CurrentEntities.get(action.Entity)!;
		if (actionEntity.CardId !== 'TB_BaconShop_8P_PlayerE') {
			return null;
		}

		const isAttackNode =
			action.Data.filter((data) => data instanceof TagChange)
				.map((data) => data as unknown as TagChange)
				.filter(
					(tag) =>
						tag.Name === (GameTag.HIGHLIGHT_ATTACKING_MINION_DURING_COMBAT as number) &&
						tag.Value === 0,
				).length > 0;
		if (!isAttackNode) {
			return null;
		}

		this.GameState.BattleResultSent = true;
		const attackAction = action.Data.filter((data) => data instanceof Action)
			.map((data) => data as unknown as Action)
			.find((act) => act.Type === (BlockType.ATTACK as number));

		let opponentPlayerId = this.StateFacade.OpponentPlayer!.PlayerId;
		if (attackAction == null) {
			let battleResult = 'tied';
			const gsPlayer = this.StateFacade.GsState!.GameState.CurrentEntities.get(
				this.StateFacade.LocalPlayer!.Id,
			)!;
			if (gsPlayer.GetTag(GameTag.PLAYSTATE) === (PlayState.LOST as number)) {
				battleResult = 'lost';
			} else if (gsPlayer.GetTag(GameTag.PLAYSTATE) === (PlayState.WON as number)) {
				battleResult = 'won';
			}

			let opponentHero: FullEntity | undefined = [...this.GameState.CurrentEntities.values()]
				.filter((data) => data.GetEffectiveController() === opponentPlayerId)
				.filter((data) => data.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number))
				.filter((data) => data.GetTag(GameTag.ZONE) === (Zone.PLAY as number))
				.find(() => true);
			let cardId = opponentHero?.CardId ?? null;
			if (cardId != null && this.isBaconGhost(cardId)) {
				const player = this.GameState.CurrentEntities.get(this.StateFacade.LocalPlayer!.Id)!;
				opponentPlayerId = player.GetTag(GameTag.NEXT_OPPONENT_PLAYER_ID);
				opponentHero = [...this.GameState.CurrentEntities.values()]
					.filter((entity) => entity.GetTag(GameTag.PLAYER_ID) === opponentPlayerId)
					.filter(
						(entity) =>
							!entity.IsBaconBartender() &&
							!entity.IsBaconGhost() &&
							!entity.IsBaconEnchantment(),
					)
					.find(() => true);
				cardId = opponentHero?.CardId ?? null;
			}

			return [
				GameEventProvider.Create(
					action.TimeStamp,
					'BATTLEGROUNDS_BATTLE_RESULT',
					() => ({
						Type: 'BATTLEGROUNDS_BATTLE_RESULT',
						Value: {
							Opponent: cardId,
							OpponentPlayerId: opponentPlayerId,
							Result: battleResult,
						},
					}),
					true,
					node,
				),
			];
		}

		const winner = this.GameState.CurrentEntities.get(attackAction.Entity)!;
		const result =
			winner.GetEffectiveController() === this.StateFacade.LocalPlayer!.PlayerId ? 'won' : 'lost';
		const damageTag = attackAction.Data.filter((data) => data instanceof TagChange)
			.map((data) => data as unknown as TagChange)
			.find((tag) => tag.Name === (GameTag.PREDAMAGE as number));
		const attackerEntityId = attackAction.Data.filter((data) => data instanceof TagChange)
			.map((data) => data as unknown as TagChange)
			.find(
				(tag) => tag.Name === (GameTag.ATTACKING as number) && tag.Value === 1,
			)!.Entity;
		const defenderEntityId = attackAction.Data.filter((data) => data instanceof TagChange)
			.map((data) => data as unknown as TagChange)
			.find(
				(tag) => tag.Name === (GameTag.DEFENDING as number) && tag.Value === 1,
			)!.Entity;
		const opponentEntityId =
			this.GameState.CurrentEntities.get(attackerEntityId)!.GetEffectiveController() ===
			this.StateFacade.LocalPlayer!.PlayerId
				? defenderEntityId
				: attackerEntityId;
		let opponentCardId = this.GameState.CurrentEntities.get(opponentEntityId)!.CardId;
		opponentPlayerId = this.GameState.CurrentEntities.get(opponentEntityId)!.GetTag(GameTag.PLAYER_ID);
		const mainPlayer = this.StateFacade.LocalPlayer!;
		if (opponentCardId === CardIds.Kelthuzad_TB_BaconShop_HERO_KelThuzad) {
			const playerEntity = [...this.GameState.CurrentEntities.values()]
				.filter((entity) => entity.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number))
				.filter((entity) => entity.GetTag(GameTag.ZONE) === (Zone.PLAY as number))
				.filter((entity) => entity.GetEffectiveController() === mainPlayer.PlayerId)
				.filter(
					(entity) =>
						!entity.IsBaconBartender() &&
						!entity.IsBaconGhost() &&
						!entity.IsBaconEnchantment(),
				)
				.sort((a, b) => a.Id - b.Id)
				.pop();
			const nextOpponentPlayerId = playerEntity?.GetTag(GameTag.NEXT_OPPONENT_PLAYER_ID);

			const nextOpponentCandidates = [...this.GameState.CurrentEntities.values()]
				.filter((entity) => entity.GetTag(GameTag.CARDTYPE) === (CardType.HERO as number))
				.filter((entity) => entity.GetTag(GameTag.PLAYER_ID) === nextOpponentPlayerId)
				.filter(
					(entity) =>
						!entity.IsBaconBartender() &&
						!entity.IsBaconGhost() &&
						!entity.IsBaconEnchantment(),
				);
			const nextOpponent = nextOpponentCandidates.length === 0 ? null : nextOpponentCandidates[0];

			opponentCardId = nextOpponent?.CardId ?? opponentCardId;
			opponentPlayerId = nextOpponentPlayerId ?? 0;
		}
		const damage = damageTag != null ? damageTag.Value : 0;

		return [
			GameEventProvider.Create(
				action.TimeStamp,
				'BATTLEGROUNDS_BATTLE_RESULT',
				() => ({
					Type: 'BATTLEGROUNDS_BATTLE_RESULT',
					Value: {
						Opponent: opponentCardId,
						OpponentPlayerId: opponentPlayerId,
						Result: result,
						Damage: damage,
					},
				}),
				true,
				node,
			),
		];
	}

	private isBaconGhost(cardId: string): boolean {
		return (
			cardId === CardIds.LadyDeathwhisper_TB_BaconShop_HERO_Deathwhisper ||
			cardId === CardIds.Kelthuzad_TB_BaconShop_HERO_KelThuzad
		);
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsBattleOverParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;
using System;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;
using HearthstoneReplays.Events.Parsers.Utils;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsBattleOverParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsBattleOverParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            // The tag change is only used as a fallback mechanism in case no battle result was sent
            // This can happen in case of draws sometimes
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && !GameState.BattleResultSent
                && node.Type == typeof(TagChange)
                && (node.Object as TagChange).Name == (int)GameTag.BOARD_VISUAL_STATE
                && (node.Object as TagChange).Value == 1;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(Action)
                && (node.Object as Action).Type == (int)BlockType.TRIGGER
                // Also modify this in trigger-sync KDA
                // To know the value, look at the ATTACK node in which a hero deals damage to another
                && (node.Object as Action).EffectIndex == 14;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            GameState.BattleResultSent = true;
            var tagChange = node.Object as TagChange;
            string opponentCardId = GameState.BgsCurrentBattleOpponent;
            int opponentPlayerId = GameState.BgsCurrentBattleOpponentPlayerId;
            var mainPlayer = StateFacade.LocalPlayer;
            if (opponentCardId == null || BgsUtils.IsBaconGhost(opponentCardId))
            {
                // Finding the one that is flagged as the player's NEXT_OPPONENT
                var playerEntity = GameState.CurrentEntities.Values
                    .Where(entity => entity.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO)
                    .Where(entity => entity.GetTag(GameTag.ZONE) == (int)Zone.PLAY)
                    .Where(entity => entity.GetEffectiveController() == mainPlayer.PlayerId)
                    .Where(entity => !entity.IsBaconBartender()
                        && !entity.IsBaconGhost()
                        && !entity.IsBaconEnchantment())
                    .OrderBy(entity => entity.Id)
                    .LastOrDefault();
                var nextOpponentPlayerId = playerEntity.GetTag(GameTag.NEXT_OPPONENT_PLAYER_ID);

                var nextOpponentCandidates = GameState.CurrentEntities.Values
                    .Where(entity => entity.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO)
                    .Where(entity => entity.GetTag(GameTag.PLAYER_ID) == nextOpponentPlayerId)
                    .Where(entity => !entity.IsBaconBartender()
                        && !entity.IsBaconGhost()
                        && !entity.IsBaconEnchantment())
                    .ToList();
                var nextOpponent = nextOpponentCandidates == null || nextOpponentCandidates.Count == 0 ? null : nextOpponentCandidates[0];

                opponentCardId = nextOpponent?.CardId;
                opponentPlayerId = nextOpponent?.GetTag(GameTag.PLAYER_ID) ?? 0;
            }

            var debug = opponentCardId == "TB_BaconShop_HERO_27" && opponentPlayerId == 9;
            return new List<GameEventProvider> { GameEventProvider.Create(
                    tagChange.TimeStamp,
                     "BATTLEGROUNDS_BATTLE_RESULT",
                    () => new GameEvent
                    {
                        Type = "BATTLEGROUNDS_BATTLE_RESULT",
                        Value = new
                        {
                            Opponent = opponentCardId,
                            OpponentPlayerId = opponentPlayerId,
                            Result = "tied"
                        }
                    },
                    true,
                    node)
                };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var action = node.Object as Action;
            var actionEntity = GameState.CurrentEntities[action.Entity];
            if (actionEntity.CardId != "TB_BaconShop_8P_PlayerE")
            {
                return null;
            }

            var isAttackNode = action.Data
                .Where(data => data.GetType() == typeof(TagChange))
                .Select(data => data as TagChange)
                .Where(tag => (tag.Name == (int)GameTag.HIGHLIGHT_ATTACKING_MINION_DURING_COMBAT && tag.Value == 0))
                .Count() > 0;
            if (!isAttackNode)
            {
                return null;
            }

            GameState.BattleResultSent = true;
            var attackAction = action.Data
                .Where(data => data.GetType() == typeof(Action))
                .Select(data => data as Action)
                .Where(act => act.Type == (int)BlockType.ATTACK)
                .FirstOrDefault();

            var opponentPlayerId = StateFacade.OpponentPlayer.PlayerId;
            if (attackAction == null)
            {
                var battleResult = "tied";
                var gsPlayer = StateFacade.GsState.GameState.CurrentEntities[StateFacade.LocalPlayer.Id];
                // If a player died because of demon health-consuming effects, there is no attack, but the battle isn't tied
                if (gsPlayer.GetTag(GameTag.PLAYSTATE) == (int)PlayState.LOST)
                {
                    battleResult = "lost";
                }
                else if (gsPlayer.GetTag(GameTag.PLAYSTATE) == (int)PlayState.WON)
                {
                    battleResult = "won";
                }

                var opponentHero = GameState.CurrentEntities.Values
                    .Where(data => data.GetEffectiveController() == opponentPlayerId)
                    .Where(data => data.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO)
                    .Where(data => data.GetTag(GameTag.ZONE) == (int)Zone.PLAY)
                    .FirstOrDefault();
                var cardId = opponentHero?.CardId;
                if (BgsUtils.IsBaconGhost(cardId))
                {
                    // Find the nexwt_opponent_id
                    var player = GameState.CurrentEntities[StateFacade.LocalPlayer.Id];
                    opponentPlayerId = player.GetTag(GameTag.NEXT_OPPONENT_PLAYER_ID);
                    opponentHero = GameState.CurrentEntities.Values
                        .Where(entity => entity.GetTag(GameTag.PLAYER_ID) == opponentPlayerId)
                        .Where(entity => !entity.IsBaconBartender()
                            && !entity.IsBaconGhost()
                            && !entity.IsBaconEnchantment())
                        .FirstOrDefault();
                    cardId = opponentHero?.CardId;
                }

                var debug = cardId == "TB_BaconShop_HERO_27" && opponentPlayerId == 9;
                return new List<GameEventProvider> { GameEventProvider.Create(
                    action.TimeStamp,
                     "BATTLEGROUNDS_BATTLE_RESULT",
                    () => new GameEvent
                    {
                        Type = "BATTLEGROUNDS_BATTLE_RESULT",
                        Value = new
                        {
                            Opponent = cardId,
                            OpponentPlayerId = opponentPlayerId,
                            Result = battleResult
                        }
                    },
                    true,
                    node)
                };
            }

            var winner = GameState.CurrentEntities[attackAction.Entity];
            var result = winner.GetEffectiveController() == StateFacade.LocalPlayer.PlayerId ? "won" : "lost";
            var damageTag = attackAction.Data
                .Where(data => data.GetType() == typeof(TagChange))
                .Select(data => data as TagChange)
                .Where(tag => tag.Name == (int)GameTag.PREDAMAGE)
                .FirstOrDefault();
            var attackerEntityId = attackAction.Data
                .Where(data => data.GetType() == typeof(TagChange))
                .Select(data => data as TagChange)
                .Where(tag => tag.Name == (int)GameTag.ATTACKING && tag.Value == 1)
                .FirstOrDefault()
                .Entity;
            var defenderEntityId = attackAction.Data
                .Where(data => data.GetType() == typeof(TagChange))
                .Select(data => data as TagChange)
                .Where(tag => tag.Name == (int)GameTag.DEFENDING && tag.Value == 1)
                .FirstOrDefault()
                .Entity;
            var opponentEntityId = GameState.CurrentEntities[attackerEntityId].GetEffectiveController() == StateFacade.LocalPlayer.PlayerId
                ? defenderEntityId
                : attackerEntityId;
            var opponentCardId = GameState.CurrentEntities[opponentEntityId].CardId;
            opponentPlayerId = GameState.CurrentEntities[opponentEntityId].GetTag(GameTag.PLAYER_ID);
            var mainPlayer = StateFacade.LocalPlayer;
            if (opponentCardId == Kelthuzad_TB_BaconShop_HERO_KelThuzad)
            {
                // Finding the one that is flagged as the player's NEXT_OPPONENT
                var playerEntity = GameState.CurrentEntities.Values
                    .Where(entity => entity.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO)
                    .Where(entity => entity.GetTag(GameTag.ZONE) == (int)Zone.PLAY)
                    .Where(entity => entity.GetEffectiveController() == mainPlayer.PlayerId)
                    .Where(entity => !entity.IsBaconBartender()
                        && !entity.IsBaconGhost()
                        && !entity.IsBaconEnchantment())
                    .OrderBy(entity => entity.Id)
                    .LastOrDefault();
                // Sometimes there is no player entity, but I don't know why
                var nextOpponentPlayerId = playerEntity?.GetTag(GameTag.NEXT_OPPONENT_PLAYER_ID);

                var nextOpponentCandidates = GameState.CurrentEntities.Values
                    .Where(entity => entity.GetTag(GameTag.CARDTYPE) == (int)CardType.HERO)
                    .Where(entity => entity.GetTag(GameTag.PLAYER_ID) == nextOpponentPlayerId)
                    .Where(entity => !entity.IsBaconBartender()
                        && !entity.IsBaconGhost()
                        && !entity.IsBaconEnchantment())
                    .ToList();
                var nextOpponent = nextOpponentCandidates == null || nextOpponentCandidates.Count == 0 ? null : nextOpponentCandidates[0];

                opponentCardId = nextOpponent?.CardId;
                opponentPlayerId = nextOpponentPlayerId ?? 0;
            }
            var damage = damageTag != null ? damageTag.Value : 0;

            return new List<GameEventProvider> { GameEventProvider.Create(
                action.TimeStamp,
                "BATTLEGROUNDS_BATTLE_RESULT",
                () => new GameEvent
                {
                    Type = "BATTLEGROUNDS_BATTLE_RESULT",
                    Value = new
                    {
                        Opponent = opponentCardId,
                        OpponentPlayerId = opponentPlayerId,
                        Result = result,
                        Damage = damage,
                    }
                },
                true,
                node)
            };
        }
    }
}
